src/Compiler/lex.fsl

```diff
@@ -615,54 +616,77 @@ rule token args skip = parse
        | _ :: _ -> errorR(Error(FSComp.SR.lexTripleQuoteInTripleQuote(), m))
        | [] -> ()
 
-       if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, LexerStringKind.InterpolatedStringFirst, m))
+       args.delimLength <- 1
+       if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, LexerStringKind.InterpolatedStringFirst, 1, m))
        else tripleQuoteString (buf, fin, m, LexerStringKind.InterpolatedStringFirst, args) skip lexbuf }
 
+ | ('$'+) '"' '"' '"'
+     { let buf, fin, m = startString args lexbuf
+
+       if lexbuf.SupportsFeature LanguageFeature.ExtendedStringInterpolation then
+           // Single quote in triple quote ok, others disallowed
+           match args.stringNest with
+           | _ :: _ -> errorR(Error(FSComp.SR.lexTripleQuoteInTripleQuote(), m))
+           | [] -> ()
+
+           args.delimLength <- lexeme lexbuf |> Seq.takeWhile (fun c -> c = '$') |> Seq.length
+           if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.ExtendedInterpolated, LexerStringKind.InterpolatedStringFirst, args.delimLength, m))
+           else extendedInterpolatedString (buf, fin, m, LexerStringKind.InterpolatedStringFirst, args) skip lexbuf
+       else
+           let result() =
+             if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, LexerStringKind.InterpolatedStringFirst, args.delimLength, m))
+             else tripleQuoteString (buf, fin, m, LexerStringKind.InterpolatedStringFirst, args) skip lexbuf
+           fail args lexbuf (FSComp.SR.lexExtendedStringInterpolationNotSupported()) (result())
```

Comment:
How does "fail" work here in terms of recovery?
Is rest of the lexing doomed after it, or can this still recover and give meaningful results for unrelated parts of the same file?

Reply:
(i.e. not just "errorR" here?)

Reply:
Tbh I don't understand it all that well myself, error recovery is somewhat coupled to the parser. But note that we continue tokenization - the `result` thing, passed to `fail`, which is simply returned after raising the error... Actually looking at it now, I don't see why I made `result` a function taking a `unit`, probably cargo-culted it from somewhere where it is needed (to preserve correct range most likely), will revisit.